certbot-apache/certbot_apache/configurator.py

```diff
@@ -942,6 +969,78 @@ def get_virtual_hosts(self):
                     vhs.append(new_vhost)
         return vhs
 
+    def get_virtual_hosts_v2(self):
+        """Returns list of virtual hosts found in the Apache configuration using
+        ParserNode interface.
+
+        :returns: List of :class:`~certbot_apache.obj.VirtualHost`
+            objects found in configuration
+        :rtype: list
+        """
+
+        vhs = []
+        vhosts = self.parser_root.find_blocks("VirtualHost", exclude=False)
+        for vhblock in vhosts:
+            vhs.append(self._create_vhost_v2(vhblock))
+        return vhs
+
+    def _create_vhost_v2(self, node):
+        """Used by get_virtual_hosts_v2 to create vhost objects using ParserNode
+        interfaces.
+
+        :param interfaces.BlockNode node: The BlockNode object of VirtualHost block
+        :returns: newly created vhost
+        :rtype: :class:`~certbot_apache.obj.VirtualHost`
+        """
+        addrs = set()
+        for param in node.parameters:
+            addrs.add(obj.Addr.fromstring(param))
+
+        is_ssl = False
+        sslengine = node.find_directives("SSLEngine")
+        if sslengine:
+            for directive in sslengine:
+                if directive.parameters[0].lower() == "on":
+                    is_ssl = True
+
+        # "SSLEngine on" might be set outside of <VirtualHost>
+        # Treat vhosts with port 443 as ssl vhosts
+        for addr in addrs:
+            if addr.get_port() == "443":
+                is_ssl = True
+
+        macro = False
+        # Check if the VirtualHost is contained in a mod_macro block
+        if node.find_ancestors("Macro"):
+            macro = True
+        vhost = obj.VirtualHost(
+            node.filepath, None, addrs, is_ssl, node.enabled, modmacro=macro, node=node
```

Comment:
We really don't need the path? I know it's now unnecessary, but I'm surprised it doesn't cause test failures in legacy checks. I wonder if it makes sense to do something like putting `USE_PARSERNODE` in `constants.py` and running all the tests with it on. Or do we think we'll have removed any code referencing it before we're ready to run all tests with that set?

Reply:
Currently the `VirtualHost` objects created by `_v2` implementation are not used further down in the execution, but merely just asserted against the list of `VirtualHost`s returned by the legacy implementation, and then stowed away. With the continued refactoring, we get to use them when `self.USE_PARSERNODE` boolean gets flipped.